File: lb2/java/ComputableArray.java
package lb2.java;

import java.util.Random;

public class ComputableArray {
    public final int[] arr; 

    private Integer _min = null;
    private Integer _minIndex = null;
    private final int _totalThreads;
    private int _finishedThreads = 0;

    public ComputableArray(int length, int threadsAmount) {
        this.arr = new int[length];
        this._totalThreads = threadsAmount;
        this.fillArray(100, 200);
    }

    public int[] minSync() {
        this._min = null;
        return this.iterate(0, this.arr.length);
    }

    public int[] minAsync() {
        this._min = null;
        this._findLowestValue();
        return new int[] { this._min, this._minIndex };
    }

    public synchronized void increaseCounter() {
        this._finishedThreads++;
    }

    public synchronized void unlock() {
        notify();
    }

    public synchronized void collectMin(int val, int index) {
        if (this._min == null || this._min > val) {
            this._min = val;
            this._minIndex = index;
        }
    }

    public int[] iterate(int start, int end) {
        int min = Integer.MAX_VALUE;
        int minIndex = 0;
        for (int i = start; i < end; i++) {
            int entry = this.arr[i];
            if (min > entry) {
                min = entry;
                minIndex = i;
            }
        }

        return new int[] { min, minIndex };
    }

    private void _findLowestValue() {
        int chunkSize = this.arr.length / this._totalThreads;
        int remainder = this.arr.length % this._totalThreads;

        for (int i = 0; i < this._totalThreads; i++) {
            int chunkOffsetSize = chunkSize + (i < remainder ? 1 : 0);
            int start = i * chunkOffsetSize;
            int end = start + chunkOffsetSize;
            var cw = new ChunkWorker(start, end, this);
            cw.start();
        }

        stutterComputations();
    }

    private synchronized void stutterComputations() {
        while(this._finishedThreads < this._totalThreads) {
            try {
                wait();
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        }

        this._finishedThreads = 0;
    }

    private void fillArray(int min, int max) {
        var rnd = new Random();

        for (int i = 0; i < this.arr.length; i++) {
            this.arr[i] = rnd.nextInt(max - min + 1) + min;
        }

        int negativeIdx = rnd.nextInt(0, this.arr.length);
        this.arr[negativeIdx] = -this.arr[negativeIdx];
    }
}

File: lb2/c#/lb2c#/ComputableArray.cs
namespace lb2 {
    class ComputableArray {
        private readonly int[] _arr;
        private readonly int _totalThreads;
        private readonly object _minValueLocker = new object();
        private readonly object _threadCountLocker = new object();

        private int _usedThreads = 0;
        private int _asyncMin;
        private int _minIndex;

        public int[] GetMinAsync() {
            int chunkSize = this._arr.Length / this._totalThreads;
            int remainder = this._arr.Length % this._totalThreads;

            for (int i = 0; i < this._totalThreads; i++) {
                int chunkOffsetSize = chunkSize + (i < remainder ? 1 : 0);
                int start = i * chunkOffsetSize;
                int end = start + chunkOffsetSize;
                var thread = new Thread(this._iterateAsync);
                thread.Start(new IterationBounds(){
                    start = start,
                    end = end,
                    cb = (int val, int currentMin) => val < currentMin ? val : null
                });
            }

            _stutterComputations();

            return new int[] { this._asyncMin, this._minIndex };
        }

        public int[] GetMinSync() {
            return _iterate(0, this._arr.Length, (x, currentMin) => {
                if (x < currentMin) {
                    return x;
                }

                return null;
            });
        }

        public ComputableArray(int length, int threadsAmount) {
            this._arr = new int[length];
            this._totalThreads = threadsAmount;
            this._fillArray(100, 200);
        }

        private void _fillArray(int min, int max) {
            var rnd = new Random();

            for (int i = 0; i < this._arr.Length; i++) {
                this._arr[i] = rnd.Next(min, max + 1);
            }

            int negativeIdx = rnd.Next(0, this._arr.Length);
            this._arr[negativeIdx] = -this._arr[negativeIdx];
        }

        private void _iterateAsync(object iterationBounds) {
            var ib = (iterationBounds as IterationBounds);
            if (ib != null) {
                int result = int.MaxValue;
                int index = -1;
                for (int i = ib.start; i < ib.end; i++) {
                    var cbResult = ib.cb(this._arr[i], result);
                    if (cbResult != null) {
                        index = i;
                        result = (int)cbResult;
                    }
                }

                lock(this._minValueLocker) {
                    this._collectMin(result, index);
                }

                lock(this._threadCountLocker) {
                    this._incrementUsedThreadsCount();
                }
            }
        }

        private int[] _iterate(int start, int end, Func<int, int, int?> cb) {
            int result = Int32.MaxValue;
            int index = -1;
            for (int i = start; i < end; i++) {
                var cbResult = cb(this._arr[i], result);
                if (cbResult != null) {
                    result = (int)cbResult;
                    index = i;
                }
            }

            return new int[] { result, index };
        }

        private void _stutterComputations() {
            lock(this._threadCountLocker) {
                while (this._totalThreads > this._usedThreads) {
                    Monitor.Wait(this._threadCountLocker);
                }
            }
        }

        private void _collectMin(int minValue, int index) {
            if (this._asyncMin > minValue) {
                this._asyncMin = minValue;
                this._minIndex = index;
            }
        }

        private void _incrementUsedThreadsCount() {
            this._usedThreads++;
            Monitor.PulseAll(this._threadCountLocker);
        }
    }
}

